File: Java/src/LowestCommonAncestorofDeepestLeaves.java
public class LowestCommonAncestorofDeepestLeaves{

    private int height;
    private TreeNode res;
    public TreeNode lcaDeepestLeaves(TreeNode root) {
        height = height(root);
        dfs(root, 1);
        return res;
    }

    private int height(TreeNode n){
        if(n == null) return 0;
        return Math.max(height(n.left), height(n.right)) + 1;
    }

    private boolean dfs(TreeNode n, int h){
        if(n == null) return false;

        if(h == height){
            res = n;
            return true;
        }

        boolean l = dfs(n.left, h+1);
        boolean r = dfs(n.right, h+1);

        if(l && r) res = n;
        return l || r;
    }

}

File: Python/LowestCommonAncestorofDeepestLeaves.py
class Solution:
    def lcaDeepestLeaves(self, root: TreeNode) -> TreeNode:
        self.h = self.height(root)

        self.dfs(root, 1)
        return self.res



    def height(self, n: TreeNode) -> int:
        if n == None:
            return 0

        l = self.height(n.left)
        r = self.height(n.right)
        return max(l, r) + 1

    def dfs(self, n: TreeNode, depth: int) -> bool:
        if n == None:
            return False
        
        if depth == self.h:
            self.res = n
            return True

        l = self.dfs(n.left, depth+1)
        r = self.dfs(n.right, depth+1)

        if l and r:
            self.res = n
        
        return l or r



    
